source/ocs/ocs-connection.ts

```diff
@@ -0,0 +1,66 @@
+import { ConnectionOptions } from '../types';
+
+export class OcsConnection {
+  options : ConnectionOptions;
+
+  constructor(url: string)
+  constructor(options : ConnectionOptions)
+  constructor(options : string | ConnectionOptions) {
+    if (options.constructor === String) {
+      options = { url: options as string };
+    }
+    this.options = options as ConnectionOptions;
+
+    if (this.options.url.lastIndexOf('/') === this.options.url.length - 1) {
+      this.options.url = this.options.url.substring(0, this.options.url.length - 1);
+    }
+  }
+
+  getHeader() {
+    const credentials = Buffer.from(`${this.options.username}:${(this.options.password ? this.options.password : '')}`).toString('base64');
+    const header = {
+      'Content-Type' : 'application/x-www-form-urlencoded',
+      'OCS-APIRequest' : true,
+      Accept: 'application/json',
+      Authorization: `Basic ${credentials}`
+    };
+
+    return header;
+  }
+
+  isValidResponse(response) : boolean {
+    return (response && response.ocs && response.ocs.meta);
+  }
+
+  request(error, response, body, callback) {
+    if (error) {
+      callback(error, null);
+      return;
+    }
+    if (response.statusCode !== 200) {
+      callback({
+        code: response.statusCode,
+        message: response.statusMessage
+      }, null);
+
+      return;
+    }
+
+    const jsonBody = JSON.parse(body);
+    if (jsonBody && jsonBody.ocs && jsonBody.ocs.meta) {
+      // Response is well-formed
+      callback(null, jsonBody.ocs);
+    } else {
+      // Server said everything's fine but response is malformed
+      callback({
+        code: 500,
```

Comment:
I'm not sure 500 is the correct error code here… what could be an example for this happening?

Reply:
This could happen is where we DO get a status 200, so clearly the server is happy, but for whatever reason we did not get the fields that we need (and should be receiving).
I don't know in what situations this could happen, all I know is that something went wrong server side.

Reply:
After discussion this will be left as 500, but only because there is no better status code available.